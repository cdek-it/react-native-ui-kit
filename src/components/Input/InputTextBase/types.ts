import type { Ref, ReactNode } from 'react'
import type { TextInputProps, ViewStyle, TextInput } from 'react-native'

/** @see TextInputProps */
export interface InputTextBaseProps
  extends Omit<TextInputProps, 'style' | 'editable' | 'secureTextEntry'> {
  /**
   * Управление отображения иконки очистки поля
   * @default true
   */
  clearable?: boolean
  /**
   * Управление режимом скрытия текста
   * - `true` — всегда скрывать (как в TextInput)
   * - `false` — не скрывать
   * - `"toggleable"` — скрытие управляется пользователем через кнопку-глаз
   */
  secureTextEntry?: boolean | 'toggleable'
  /** Озвучка для кнопки очистки поля */
  clearButtonAccessibilityLabel?: string
  /** Управление стилем контейнера поля ввода */
  containerStyle?: ViewStyle
  /** Управление доступностью поля */
  disabled?: boolean
  /** Ref для управления полем ввода */
  inputRef?: Ref<TextInput | null>
  // Функция для рендера поля ввода.
  // Используется, когда необходимо использовать отличный от стандартного компонент.
  // Например, для реализации масок
  //
  renderTextInput?: (props: RenderTextInputArgs) => ReactNode
  /** Управление состоянием компонента */
  state?: 'default' | 'danger'
}

export type RenderTextInputArgs = TextInputProps & { inputRef: Ref<TextInput> }
