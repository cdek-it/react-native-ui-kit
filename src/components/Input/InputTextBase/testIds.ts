export const InputTextBaseTestId = {
  default: 'InputTextBase',
  focusOutline: 'FocusOutline',
  dangerOutline: 'DangerOutline',
  loading: 'Loading',
  clearButton: 'ClearButton',
  secureInputButton: 'SecureInputButton',
  disabledIcon: 'DisabledIcon',
}
