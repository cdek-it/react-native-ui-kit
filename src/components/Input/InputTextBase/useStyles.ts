import { makeStyles } from '../../../utils/makeStyles'

export const useStyles = makeStyles(({ theme }) => ({
  container: {
    minHeight: 35,
    flexDirection: 'row',
    borderWidth: 1,
    borderRadius: theme.General.borderRadius,
    borderColor: theme.Form.InputText.inputBorderColor,
    backgroundColor: theme.Form.InputText.inputBg,
  },
  danger: { borderColor: theme.Form.InputText.inputErrorBorderColor },
  disabled: {
    opacity: 0.6,
    borderColor: theme.Form.InputText.inputBorderColor,
    backgroundColor: theme.Button.Disabled.disabledButtonBg,
  },
  input: {
    flex: 1,
    paddingVertical: theme.Form.InputText.inputPaddingTopBottom,
    paddingHorizontal: theme.Form.InputText.inputPaddingLeftRight,
    borderRadius: theme.General.borderRadius,
    fontSize: 14,
    color: theme.Form.InputText.inputTextColor,
    overflow: 'hidden',
    includeFontPadding: false,
    verticalAlign: 'middle',
  },
  placeholderTextColor: {
    color: theme.Form.InputText.inputPlaceholderTextColor,
  },
  outline: { position: 'absolute' },
  outlineWidth: { borderWidth: theme.General.focusShadowWidth },
  focusOutline: { backgroundColor: theme.General.focusOutlineColor },
  dangerOutline: { backgroundColor: theme.General.focusOutlineErrorColor },
  rightContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    paddingRight: theme.Form.InputText.inputPaddingLeftRight,
    maxHeight: 54,
    gap: theme.Form.InputText.inputPaddingLeftRight,
    overflow: 'hidden',
  },
  rightIcon: { color: theme.Form.InputText.inputIconColor },
  iconSize: { width: 14, height: 14 },
}))
