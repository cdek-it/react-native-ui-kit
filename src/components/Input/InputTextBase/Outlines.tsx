import { memo, type RefObject, useImperativeHandle, useMemo } from 'react'
import type { ViewStyle } from 'react-native'

import Animated, { useSharedValue, withTiming } from 'react-native-reanimated'

import { makeStyles } from '../../../utils/makeStyles'

import { InputTextBaseTestId } from './testIds'

export interface OutlinesRef {
  focus: () => void
  blur: () => void
  setDanger: (on: boolean) => void
}

interface OutlinesProps {
  readonly containerStyle?: ViewStyle
  readonly baseBorderRadius?: number
  readonly disabled?: boolean
  readonly makeTestId: (name: string) => string
  readonly outlinesRef: RefObject<OutlinesRef>
}

const withOutlineAnimation = (toValue: number) =>
  withTiming(toValue, { duration: 100 })

export const Outlines = memo<OutlinesProps>(
  ({ disabled, makeTestId, containerStyle, baseBorderRadius, outlinesRef }) => {
    const styles = useStyles()
    const focusOutlineWidth = useSharedValue(0)
    const dangerOutlineWidth = useSharedValue(0)

    const focus = () => {
      focusOutlineWidth.value = withOutlineAnimation(
        -styles.outlineWidth.borderWidth
      )
    }

    const blur = () => {
      focusOutlineWidth.value = withOutlineAnimation(0)
    }

    const setDanger = (value: boolean) => {
      dangerOutlineWidth.value = withOutlineAnimation(
        value ? -styles.outlineWidth.borderWidth : 0
      )
    }

    useImperativeHandle(outlinesRef, () => ({ focus, blur, setDanger }))

    const focusOutlineAnimatedStyles = useMemo(() => {
      return {
        top: focusOutlineWidth,
        right: focusOutlineWidth,
        bottom: focusOutlineWidth,
        left: focusOutlineWidth,
      }
    }, [focusOutlineWidth])

    const dangerOutlineAnimatedStyles = useMemo(
      () => ({
        top: dangerOutlineWidth,
        right: dangerOutlineWidth,
        bottom: dangerOutlineWidth,
        left: dangerOutlineWidth,
      }),
      [dangerOutlineWidth]
    )

    const outlineStyles = useMemo<ViewStyle>(
      () => ({
        borderRadius: containerStyle?.borderRadius ?? baseBorderRadius,
        borderTopRightRadius: containerStyle?.borderTopRightRadius,
        borderBottomRightRadius: containerStyle?.borderBottomRightRadius,
        borderBottomLeftRadius: containerStyle?.borderBottomLeftRadius,
        borderTopLeftRadius: containerStyle?.borderTopLeftRadius,
      }),
      [
        containerStyle?.borderBottomLeftRadius,
        containerStyle?.borderBottomRightRadius,
        containerStyle?.borderRadius,
        containerStyle?.borderTopLeftRadius,
        containerStyle?.borderTopRightRadius,
        baseBorderRadius,
      ]
    )

    return disabled ? null : (
      <>
        <Animated.View
          style={[
            styles.outline,
            outlineStyles,
            styles.focusOutline,
            focusOutlineAnimatedStyles,
          ]}
          testID={makeTestId(InputTextBaseTestId.focusOutline)}
        />
        <Animated.View
          style={[
            styles.outline,
            outlineStyles,
            styles.dangerOutline,
            dangerOutlineAnimatedStyles,
          ]}
          testID={makeTestId(InputTextBaseTestId.dangerOutline)}
        />
      </>
    )
  }
)

export const useStyles = makeStyles(({ theme }) => ({
  outline: { position: 'absolute' },
  outlineWidth: { borderWidth: theme.General.focusShadowWidth },
  focusOutline: { backgroundColor: theme.General.focusOutlineColor },
  dangerOutline: { backgroundColor: theme.General.focusOutlineErrorColor },
}))
