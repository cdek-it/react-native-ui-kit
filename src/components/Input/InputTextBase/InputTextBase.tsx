import {
  IconEye,
  IconEyeOff,
  IconLoader2,
  IconLock,
  IconX,
} from '@tabler/icons-react-native'
import {
  memo,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef,
  useState,
} from 'react'
import {
  TextInput,
  View,
  type TextInputFocusEventData,
  type NativeSyntheticEvent,
  TouchableOpacity,
  type ViewStyle,
} from 'react-native'
import Animated from 'react-native-reanimated'

import { useLoadingRotationAnimation } from '../../../hooks/useLoadingRotationAnimation'
import { useMakeTestId } from '../../../hooks/useMakeTestId'

import { Outlines, type OutlinesRef } from './Outlines'
import { InputTextBaseTestId } from './testIds'
import type { InputTextBaseProps, RenderTextInputArgs } from './types'
import { useStyles } from './useStyles'

interface PrivateInputTextBaseProps {
  inputStyle?: ViewStyle
  loading?: boolean
}

/**
 * Базовое поле
 * @link https://www.figma.com/design/4TYeki0MDLhfPGJstbIicf/UI-kit-PrimeFace-(DS)?node-id=484-5470&m=dev
 * @see InputText
 */
export const InputTextBase = memo<
  InputTextBaseProps & PrivateInputTextBaseProps
>(
  ({
    state,
    clearable = true,
    secureTextEntry: secureTextEntryProp = false,
    inputRef: propsInputRef,
    disabled,
    containerStyle,
    inputStyle,
    loading,
    renderTextInput,
    clearButtonAccessibilityLabel,
    ...otherProps
  }) => {
    const styles = useStyles()
    const inputRef = useRef<TextInput>(null)
    const outlineRef = useRef<OutlinesRef>(null)
    const [valueState, setValueState] = useState('')

    const onFocus = useCallback(
      (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
        outlineRef.current?.focus()
        otherProps.onFocus?.(e)
      },
      [otherProps]
    )

    const onBlur = useCallback(
      (e: NativeSyntheticEvent<TextInputFocusEventData>) => {
        outlineRef.current?.blur()
        otherProps.onBlur?.(e)
      },
      [otherProps]
    )

    useEffect(() => {
      outlineRef.current?.setDanger(state === 'danger')
    }, [state])

    const onChangeText = useCallback(
      (nextValue: string) => {
        otherProps.onChangeText?.(nextValue)
        setValueState(nextValue)
      },
      [otherProps]
    )

    const clear = useCallback(() => {
      inputRef.current?.clear()
      onChangeText('')
    }, [onChangeText])

    const value = useMemo(
      () => otherProps.value ?? valueState,
      [otherProps.value, valueState]
    )

    const showClearButton = useMemo(
      () => clearable && !!value.length,
      [clearable, value.length]
    )

    const loadingAnimatedStyle = useLoadingRotationAnimation(loading)

    useImperativeHandle(propsInputRef, () => inputRef.current)

    const { makeTestId } = useMakeTestId(
      otherProps.testID || InputTextBaseTestId.default
    )

    const [userDefinedSecureTextEntry, setUserDefinedSecureTextEntry] =
      useState(true)
    const secureTextEntry = useMemo(
      () =>
        secureTextEntryProp === 'toggleable'
          ? userDefinedSecureTextEntry
          : secureTextEntryProp,
      [secureTextEntryProp, userDefinedSecureTextEntry]
    )
    const toggleUserDefinedSecureTextEntry = useCallback(
      () => setUserDefinedSecureTextEntry((old) => !old),
      []
    )

    const texInputProps = useMemo<RenderTextInputArgs>(
      () => ({
        placeholderTextColor: styles.placeholderTextColor.color,
        ...otherProps,
        testID: makeTestId(),
        editable: !disabled,
        secureTextEntry,
        style: [styles.input, inputStyle],
        inputRef,
        value,
        onBlur,
        onChangeText,
        onFocus,
      }),
      [
        disabled,
        secureTextEntry,
        inputStyle,
        makeTestId,
        onBlur,
        onChangeText,
        onFocus,
        otherProps,
        styles.input,
        styles.placeholderTextColor.color,
        value,
      ]
    )

    return (
      <>
        <Outlines
          baseBorderRadius={styles.container.borderRadius}
          containerStyle={containerStyle}
          disabled={disabled}
          makeTestId={makeTestId}
          outlinesRef={outlineRef}
        />
        <View
          style={[
            styles.container,
            containerStyle,
            state === 'danger' && styles.danger,
            disabled && styles.disabled,
          ]}
        >
          {renderTextInput ? (
            renderTextInput(texInputProps)
          ) : (
            <TextInput {...texInputProps} ref={inputRef} />
          )}

          <View style={styles.rightContainer}>
            {loading ? (
              <Animated.View
                style={loadingAnimatedStyle}
                testID={makeTestId(InputTextBaseTestId.loading)}
              >
                <IconLoader2
                  color={styles.rightIcon.color}
                  height={styles.iconSize.height}
                  width={styles.iconSize.width}
                />
              </Animated.View>
            ) : null}

            {showClearButton && !disabled ? (
              <TouchableOpacity
                accessibilityLabel={clearButtonAccessibilityLabel}
                testID={makeTestId(InputTextBaseTestId.clearButton)}
                onPress={clear}
              >
                <IconX
                  color={styles.rightIcon.color}
                  height={styles.iconSize.height}
                  width={styles.iconSize.width}
                />
              </TouchableOpacity>
            ) : null}

            {secureTextEntryProp === 'toggleable' ? (
              <TouchableOpacity
                testID={makeTestId(InputTextBaseTestId.secureInputButton)}
                onPress={toggleUserDefinedSecureTextEntry}
              >
                {userDefinedSecureTextEntry ? (
                  <IconEye
                    color={styles.rightIcon.color}
                    height={styles.iconSize.height}
                    width={styles.iconSize.width}
                  />
                ) : (
                  <IconEyeOff
                    color={styles.rightIcon.color}
                    height={styles.iconSize.height}
                    width={styles.iconSize.width}
                  />
                )}
              </TouchableOpacity>
            ) : null}

            {disabled ? (
              <IconLock
                color={styles.rightIcon.color}
                height={styles.iconSize.height}
                testID={makeTestId(InputTextBaseTestId.disabledIcon)}
                width={styles.iconSize.width}
              />
            ) : null}
          </View>
        </View>
      </>
    )
  }
)
