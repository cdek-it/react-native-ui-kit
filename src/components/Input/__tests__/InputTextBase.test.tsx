import { fireEvent, render, userEvent } from '@testing-library/react-native'

import { TextInput } from 'react-native-gesture-handler'

import { InputTextBase } from '../InputTextBase/InputTextBase'
import type { RenderTextInputArgs } from '../InputTextBase/types'

describe('InputTextBase component functionality tests', () => {
  test('should render outline elements', () => {
    const { getByTestId } = render(<InputTextBase />)

    expect(getByTestId('InputTextBaseDangerOutline')).toBeOnTheScreen()
    expect(getByTestId('InputTextBaseFocusOutline')).toBeOnTheScreen()
  })

  test('should NOT be render outline elements in disabled state', () => {
    const { queryByTestId } = render(<InputTextBase disabled />)

    expect(queryByTestId('InputTextBaseDangerOutline')).not.toBeOnTheScreen()
    expect(queryByTestId('InputTextBaseFocusOutline')).not.toBeOnTheScreen()
  })

  test('should set editable prop correctly', () => {
    const { queryByTestId, update } = render(<InputTextBase />)

    expect(queryByTestId('InputTextBase')).toHaveProp('editable', true)

    update(<InputTextBase disabled />)

    expect(queryByTestId('InputTextBase')).toHaveProp('editable', false)
  })

  test('should handle focus event', () => {
    const onFocusMock = jest.fn()
    const { queryByTestId } = render(<InputTextBase onFocus={onFocusMock} />)
    const input = queryByTestId('InputTextBase')

    fireEvent(input, 'focus')

    expect(onFocusMock).toHaveBeenCalled()
  })

  test('should handle blur event', () => {
    const onBlurMock = jest.fn()
    const { queryByTestId } = render(<InputTextBase onBlur={onBlurMock} />)
    const input = queryByTestId('InputTextBase')

    fireEvent(input, 'blur')

    expect(onBlurMock).toHaveBeenCalled()
  })

  test('should handle text change', () => {
    const onChangeTextMock = jest.fn()
    const { queryByTestId } = render(
      <InputTextBase onChangeText={onChangeTextMock} />
    )
    const input = queryByTestId('InputTextBase')

    fireEvent.changeText(input, 'new text')

    expect(onChangeTextMock).toHaveBeenCalledWith('new text')
  })

  test('should show clear button only when value exists', async () => {
    const { queryByTestId } = render(<InputTextBase />)
    const input = queryByTestId('InputTextBase')
    const user = userEvent.setup()
    let clearButton = queryByTestId('InputTextBaseClearButton')

    expect(clearButton).not.toBeOnTheScreen()

    await user.type(input, 'text')
    clearButton = queryByTestId('InputTextBaseClearButton')

    expect(clearButton).toBeOnTheScreen()
  })

  test('should NOT show clear button when input is disabled', async () => {
    const { queryByTestId } = render(<InputTextBase disabled value='text' />)
    const clearButton = queryByTestId('InputTextBaseClearButton')

    expect(clearButton).not.toBeOnTheScreen()
  })

  test('should clear text when clear button is pressed', async () => {
    const { queryByTestId } = render(<InputTextBase />)
    const input = queryByTestId('InputTextBase')
    const user = userEvent.setup()

    await user.type(input, 'text')

    const clearButton = queryByTestId('InputTextBaseClearButton')

    expect(input).toHaveProp('value', 'text')

    fireEvent.press(clearButton)

    expect(input).toHaveProp('value', '')
  })

  test('should show disabled icon when input is disabled', () => {
    const { queryAllByTestId, update } = render(<InputTextBase />)
    let [disabledIcon] = queryAllByTestId('InputTextBaseDisabledIcon')

    expect(disabledIcon).toBeFalsy()

    update(<InputTextBase disabled />)
    disabledIcon = queryAllByTestId('InputTextBaseDisabledIcon')

    expect(disabledIcon).toBeDefined()
  })

  test('should show loading icon', () => {
    const { queryByTestId, update } = render(<InputTextBase />)
    let loadingIcon = queryByTestId('InputTextBaseLoading')

    expect(loadingIcon).not.toBeOnTheScreen()

    update(<InputTextBase loading />)
    loadingIcon = queryByTestId('InputTextBaseLoading')

    expect(loadingIcon).toBeOnTheScreen()
    expect(loadingIcon).toHaveAnimatedStyle({ transform: [{ rotate: '0deg' }] })
  })

  test('should pass args to custom render function correctly', () => {
    const renderTextInput = jest.fn()

    render(<InputTextBase renderTextInput={renderTextInput} />)

    expect(renderTextInput).toHaveBeenCalled()
    expect(renderTextInput.mock.calls[0]).toMatchSnapshot()
  })

  test('should set clear button accessibility label', () => {
    const renderTextInput = (props: RenderTextInputArgs) => (
      <TextInput {...props} testID='MyTextInput' />
    )

    const { getByTestId } = render(
      <InputTextBase
        clearable
        clearButtonAccessibilityLabel='Clear'
        renderTextInput={renderTextInput}
      />
    )
    const input = getByTestId('MyTextInput')
    fireEvent.changeText(input, 'ку')

    const clearButton = getByTestId('InputTextBaseClearButton')

    expect(clearButton.props.accessibilityLabel).toBe('Clear')
  })

  test('should render secured field when secureTextEntry=true', () => {
    const { getByTestId } = render(<InputTextBase secureTextEntry />)

    expect(getByTestId('InputTextBase')).toHaveProp('secureTextEntry', true)
  })

  test('should show eye icon when secureTextEntry=toggleable', () => {
    const { queryByTestId, getByTestId, update } = render(<InputTextBase />)

    // Проверяем что если secureTextEntry не задан, нет кнопки с глазом
    expect(
      queryByTestId('InputTextBaseSecureInputButton')
    ).not.toBeOnTheScreen()

    // Проверяем что если secureTextEntry='toggleable' кнопка с глазом выводится
    update(<InputTextBase secureTextEntry='toggleable' />)

    const secureInputButton = getByTestId('InputTextBaseSecureInputButton')

    expect(secureInputButton).toBeOnTheScreen()

    // Проверяем что у инпута secureTextEntry=true
    const input = getByTestId('InputTextBase')

    expect(input).toHaveProp('secureTextEntry', true)

    // Нажимаем на кнопку с глазом
    fireEvent.press(secureInputButton)

    // Проверяем что у инпута secureTextEntry=false
    expect(input).toHaveProp('secureTextEntry', false)
  })
})
