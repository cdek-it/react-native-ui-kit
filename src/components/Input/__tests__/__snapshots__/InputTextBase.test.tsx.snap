// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputTextBase component functionality tests should pass args to custom render function correctly 1`] = `
[
  {
    "editable": true,
    "inputRef": {
      "current": null,
    },
    "onBlur": [Function],
    "onChangeText": [Function],
    "onFocus": [Function],
    "placeholderTextColor": "rgba(0, 0, 0, 0.6000)",
    "secureTextEntry": false,
    "style": [
      {
        "borderRadius": 7,
        "color": "rgba(0, 0, 0, 0.8000)",
        "flex": 1,
        "fontSize": 14,
        "includeFontPadding": false,
        "overflow": "hidden",
        "paddingHorizontal": 10.5,
        "paddingVertical": 0,
        "verticalAlign": "middle",
      },
      undefined,
    ],
    "testID": "InputTextBase",
    "value": "",
  },
]
`;
